### PR TITLE
Add optional serde serialization to all types

### PR DESCRIPTION
**Pull request by kamalmarhubi**
This derives serde::{Serialize, Deserialize} on all types, gated by the
already-existing serialize-serde feature. The rustc-serialize
serialization is moved to a separate serialize-rustc feature, which is
enabled by default for compatibility.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -19,6 +19,8 @@ version = "0.4.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
  "rustc-serialize 0.3.24 (registry+https://github.com/rust-lang/crates.io-index)",
+ "serde 1.0.11 (registry+https://github.com/rust-lang/crates.io-index)",
+ "serde_derive 1.0.11 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (nrc) [inline Cargo.lock:23]: The Cargo.lock changes don't belong in this PR, I think.

File: src/lib.rs [hunk 2]
```diff
@@ -26,7 +27,9 @@ use std::path::PathBuf;
 use config::Config;
 
 
-#[derive(Debug, Clone, RustcDecodable, RustcEncodable)]
+#[cfg_attr(feature = "serialize-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serialize-rustc", derive(RustcDecodable, RustcEncodable))]
+#[derive(Debug, Clone, Default)]
```
Comment (nrc) [inline src/lib.rs]: Why add `Default` here?

**Review by nrc — COMMENTED** [review]
Thanks for the PR, this looks like a sensible thing to add and the comments inline are just for minor things.

What is your goal with the changes?

File: Cargo.lock (hunk 1 above)
Comment (kamalmarhubi) [reply Cargo.lock:23]: I think they do, as the addition of the optional dependency on the feature means cargo needs to know about those dependencies, where it didn't before. I'm a bit unsure on this, so let me know if I'm wrong.

File: src/lib.rs (hunk 2 above)
Comment (kamalmarhubi) [reply src/lib.rs]: Removed.

**Comment by kamalmarhubi** [comment]
Regarding use: I am working on using `-Zsave-analysis` data, and would prefer to use serde. Specifically, I was working on revamping the [Rust stuff in Kythe](https://github.com/google/kythe/tree/master/kythe/rust) to use output from `-Zsave-analysis` where possible. It was added as a custom lint pass in a Google Summer of Code project back in 2016. The churn in compiler internals since summer 2016 is kind of huge, so decoupling from it to a more stable (but not Stable™) interface will be a huge improvement in ongoing Rust support in Kythe.

It may or may not be useful for the RLS project to add Serde support, which I leave up to you. :-)

File: Cargo.lock (hunk 1 above)
Comment (nrc) [reply Cargo.lock:23]: 👍

**Comment by nrc** [comment]
Thanks! Merging since this seems like it might be generally useful. However, note that this won't work as intended - the outputs of Serde and rustc_serialize are not compatible so you can't deserialise the output of save-analysis using Serde. IIRC, their encoding of some enums is different.

I would recommend that Kythe use rls-analysis rather than rls-data - it provides an API to the data and is intended to be more stable. Not sure if you want exactly Kythe or something like Kythe, if the latter, then you should look at [cargo src (rustw)](https://github.com/nrc/rustw) - it's currently a bit of a mess, but is slowly turning into a good tool.
